fix tab-extension ipc layer and add dependency injection back

MattWiethoff opened the pull request:
Makes the tab-to-extension ipc layer follow the same pattern as the
plugin-to-client:
 - Dependency injection to avoid passing state around.
 - Moves ipc implementation details out of command implementation
 - Fixes multiple calls to promises.
 - Renamed modules to be more descriptive.

Fixes:
- "click" incorrectly showing overlay
- dictate command taking forever on empty tab
- getting empty editor states after tab close

MattWiethoff committed:
fix tab-extension ipc layer and add dependency injection back

Makes the tab-to-extension ipc layer follow the same pattern as the
plugin-to-client:
 - Dependency injection to avoid passing state around.
 - Moves ipc implementation details out of command implementation
 - Fixes multiple calls to promises.
 - Renamed modules to be more descriptive.

Fixes:
- "click" incorrectly showing overlay
- dictate command taking forever on empty tab
- getting empty editor states after tab close

tmacwill reviewed (commented):
read through, and this looks good. I think at some point we want to use a real typescript dependency injection framework rather than making things static / initializing them in the right order, but out of scope here.

leaving for @sam-dixon to review + approve!

sam-dixon reviewed (approved):
Thanks for doing this! Looks good to me.

sam-dixon committed:
add summary elements to show links

sam-dixon committed:
Revert "add summary elements to show links"

This reverts commit e7dc307b06b0572c00feca30a60ead0f8c9e3cd3.

sam-dixon committed:
add select elements to show links overlay

sam-dixon committed:
add codemirror editor instances to list of inputs

sam-dixon committed:
fix default file name in codemirror instances

MattWiethoff committed:
fix clickables error on empty page

Went through the rpc calls that expected return values and made sure
they did okay things when the empty tab causes them to error out. This
ended up just being the clickables rpc call. Removed the paste one
because that codepath is no longer being hit.

Also converted promises to async.

sam-dixon committed:
Address bar type + use bugfix (#23)

* add event listeners for when page loses focus (focus on address bar)

* use hasFocus

* use filename to set if address bar is active, set error state in extension

Co-authored-by: Sam Dixon <sam@serenade.ai>

sam-dixon committed:
change address bar filename to empty string

MattWiethoff committed:
fix errors going through null pointer exception

Behavior ended up being the right thing, but only because a null pointer
exceptionw as thrown.
